README.md
# Passerelle de paiement Axepta pour WooCommerce

Accepter les paiements par carte bancaire sur votre boutique WooCommerce en utilisant la passerelle de paiement Axepta.

## Configuration requise

* PHP : 7.3 et plus
* WordPress : 5.6.9 et plus
* WooCommerce : 5.8 et plus
* ModRewrite sur votre serveur Apache
* Un compte Axepta - BNP Paribas avec un environnement 3DS2 acceptant les appels à Platform HTML Forms (paySSL)

Il se peut que le plugin fonctionne sur des versions antérieures de PHP, WordPress ou WooCommerce, mais il n'a pas été testé dans d'autres conditions que celles mentionnées ci-dessus.

## Installation & mises à jour

La première installation est manuelle. [Téléchargez l'archive ici](https://depot.studiomaiis.net/wordpress/woocommerce-gateway-axepta.zip) et installez le plugin de manière standard :
* via FTP une fois l'archive décompressée dans `wp-content/plugins/`,
* ou via la page d'administration des extensions.

Si vous téléchargez le code depuis Github, renommer le répertoire `wp-plugin-woocommerce-gateway-axepta(*)` en `woocommerce-gateway-axepta`.

Une fois le module installé, les mises à jour se font depuis le back-office de WordPress.

## Questions fréquentes

### A qui s'adresse ce plugin ?

Aux développeurs et aux intégrateurs.

### Ce plugin supporte-t'il les remboursements ?

Non. Il existe d'autres plugins (payants) qui proposent les remboursements. Aucune affiliation avec les auteurs de ces plugins.

### Ce plugin supporte-t'il les paiements récurrents, les abonnements ?

Non. Il existe d'autres plugins (payants) qui proposent des paiements récurrents. Aucune affiliation avec les auteurs de ces plugins.

### Est-ce qu'un certificat SSL est nécessaire ?

Non, ce n'est pas nécessaire, mais c'est hautement recommandé pour ajouter un niveau supplémentaire de sécurité pour vos clients. Techniquement, à la validation d'une commande, les utilisateurs sont redirigés vers les serveurs *Axepta* via une requête HTTPS.

### Ce plugin supporte-til les environnements de test et de production ?

Oui, les réglages de ces 2 modes sont stockés en base et le passage de l'un à l'autre se fait par le biais d'une case à cocher dans l'administration du mode de paiement. Les identifiants de test sont ceux de l'environnement générique *Axepta*. Ces derniers sont en lecture seule car les identifiants de votre environnement de test dédié (dont le MID se termine par `_t`) ne fonctionnent actuellement pas en 3DS2 ([cf. documentation](https://docs.axepta.bnpparibas/display/DOCBNP/Test+modes))

### Comment puis-je configurer ce module de paiement ?

Si vous avez le MerchantID (ou MID dans la doc), la clé HMAC et la clé de chiffrement Blowfish, vous avez tout ce qu'il vous faut et vous n'aurez aucun mal à les saisir aux endroits réservés !

![Configuration de la passerelle Axepta dans WooCommerce](https://depot.studiomaiis.net/screenshots/woo_axepta_config.png "Configuration de la passerelle Axepta dans WooCommerce")

Veuillez vous référer à la [documentation officielle](https://docs.axepta.bnpparibas/display/DOCBNP/Premiers+pas+avec+AXEPTA+BNP+Paribas) pour récupérer vos identifiants.

Attention à ne pas utiliser les tickets Github pour des demandes de support.

### Sur l'environnement de test, où puis-je trouver les cartes de test ?

Les cartes de test peuvent être trouvées [ici](https://docs.axepta.bnpparibas/display/DOCBNP/Test+Cards+-+Authentication).

### Un support, une aide sont-ils proposés pour récupérer les informations MerchantId, clés HMAC et Blowfish ?

Non, ce plugin s'adresse plutôt à des développeurs et intégrateur qui savent récupérer ces informations. Veuillez vous référer à la [documentation officielle](https://docs.axepta.bnpparibas/display/DOCBNP/Premiers+pas+avec+AXEPTA+BNP+Paribas) ou contacter le support *Axepta*.

Cherchez de l'aide dans la [section discussion de ce projet Github](https://github.com/studiomaiis/wp-plugin-woocommerce-gateway-axepta/discussions).

Si vraiment vous ne savez pas faire, vous pouvez m'envoyer une demande sur mon [site internet](https://www.studiomaiis.net), toute intervention fera l'objet d'un devis et d'une facture. Sachez que récupérer ces informations peut s'avérer assez chronophage.

### Comment passe-t'on en mode production ?

Lorsque vous avez récupéré vos identifiants de production MerchantID, les clés HMAC et Blowfish, saisissez-les dans la configuration du prestataire et activez-le.

### J'ai trouvé un bug, comment puis-je le faire remonter ?

Séquence à suivre :
1. Vérifiez que vous avez la configuration requise : PHP, WordPress et WooCommerce,
2. Vérifiez que votre serveur est accessible - exemple : si vous développez en local, les serveurs Axepta ne pourront pas envoyer leurs réponses à votre serveur,
3. Activez les logs et visualisez leurs contenus,
4. Testez avec le thème Storefront et désactivez les plugins qui pourraient interférer,
5. Regardez dans la [section discussion de ce projet](https://github.com/studiomaiis/wp-plugin-woocommerce-gateway-axepta/discussions) pour voir si d'autres utilisateurs n'ont pas eu et/ou résolu le même problème,
6. Contactez le support *Axepta*, ils peuvent avoir des informations additionnelles,
7. Soumettez un bug sur [GitHub](https://github.com/studiomaiis/wp-plugin-woocommerce-gateway-axepta/issues).

S'il s'agit d'une demande de support, vous pouvez m'envoyer une demande sur mon [site internet](https://www.studiomaiis.net), toute intervention fera l'objet d'un devis et d'une facture. 



CHANGELOG.md
#### Version 0.2

* Tests de mise à jour automatique

#### Version 0.1

* Premiers tests



includes/class-axepta-payssl.php
<?php

if ( ! defined( 'ABSPATH' ) ) {
	exit;
}

class Axepta_PaySSL {
	
	public const PAYSSL_URL = 'https://paymentpage.axepta.bnpparibas/payssl.aspx';
	
	public $supported_currencies = array(
		'EUR',
		'USD',
		'CHF',
		'GBP',
		'CAD',
		'JPY',
		'MXP',
		'TRY',
		'AUD',
		'NZD',
		'NOK',
		'BRC',
		'ARP',
		'KHR',
		'TWD',
		'SEK',
		'DKK',
		'KRW',
		'SGD',
		'XPF',
		'XOF',
	);
	
	private $supported_languages = array(
		'de',
		'en',
		'fr',
		'it',
		'pt',
		'es',
		'nl',
	);

	private $form_action = self::PAYSSL_URL;
	
	private $merchant_id;
	private $hmac_key;
	private $blowfish_key;
	
	private $message_version = '2.0';
	private $transaction_id;
	private $reference_number;
	private $amount;
	private $amount_in_cents;
	private $currency;
	private $order_description;
	private $url_notify;
	private $url_back;
	private $url_success;
	private $url_failure;
	private $user_data;
	private $response = 'encrypt';
	private $language;
	
	private $hmac_signature = '';
	
	private $blowfish_parameters = array();
	private $blowfish_mandatory_request_parameters = array( 
		'MsgVer',
		'TransID',
		'Amount',
		'Currency',
		'RefNr',
		'URLNotify',
		'URLBack',
		'URLSuccess',
		'URLFailure',
		'OrderDesc',
		'UserData',
		'Response',
	);
	private $blowfish_mandatory_response_parameters = array( 
		'PayID',
		'TransID',
		'mid',
		'Status',
		'Code',
	);
	private $blowfish_data = '';
	private $blowfish_length = '';
	
	// private $response_data = '';
	// private $response_seal = '';
	
	private $response_parameters = array();
	private $response_is_valid = false;
	private $response_code = '';
	private $response_is_successful = false;


	public function __construct() { }

	
	public function get_supported_currencies() {
		return $this->supported_currencies;
	}


	public function get_supported_languages() {
		return $this->supported_languages;
	}


	public function set_merchant_id( $merchant_id ) {
		$this->merchant_id = $merchant_id;
		$this->blowfish_parameters[ 'MerchantID' ] = $merchant_id;
	}


	public function set_hmac_key( $hmac_key ) {
		$this->hmac_key = $hmac_key;
	}


	public function set_blowfish_key( $blowfish_key ) {
		$this->blowfish_key = $blowfish_key;
	}


	public function set_transaction_id( $transaction_id ) {
		$this->transaction_id = $transaction_id;
		$this->blowfish_parameters[ 'TransID' ] = $transaction_id;
		$this->order_description = "Transaction $transaction_id";
		if ($this->merchant_id == 'BNP_DEMO_AXEPTA') {
			$this->order_description = 'Test:0000';
		}
		$this->blowfish_parameters[ 'OrderDesc' ] = $this->order_description;
	}


	public function set_reference_number( $reference_number ) {
		$this->reference_number = $reference_number;
		$this->blowfish_parameters[ 'RefNr' ] = $reference_number;
		$this->user_data = $reference_number;
		$this->blowfish_parameters[ 'UserData' ] = $reference_number;
	}


	public function set_amount( $amount ) {
		$this->amount = $amount;
		$this->amount_in_cents = intval( floatval( $amount ) * 100 );
		$this->blowfish_parameters[ 'Amount' ] = $this->amount_in_cents;
	}


	public function set_currency( $currency ) {
		if ( ! in_array( $currency, $this->supported_currencies ) ) {
			throw new Exception( "Currency not supported" );
		}
		$this->currency = $currency;
		$this->blowfish_parameters[ 'Currency' ] = $currency;
	}


	public function set_url_notify( $url_notify ) {
		$this->url_notify = $url_notify;
		$this->blowfish_parameters[ 'URLNotify' ] = $url_notify;
	}


	public function set_url_back( $url_back ) {
		$this->url_back = $url_back;
		$this->blowfish_parameters[ 'URLBack' ] = $url_back;
	}


	public function set_url_success( $url_success ) {
		$this->url_success = $url_success;
		$this->blowfish_parameters[ 'URLSuccess' ] = $url_success;
	}


	public function set_url_failure( $url_failure ) {
		$this->url_failure = $url_failure;
		$this->blowfish_parameters[ 'URLFailure' ] = $url_failure;
	}


	public function set_language( $language ) {
		if ( in_array( $language, $this->supported_languages ) ) {
			$this->language = $language;
		}
	}


	public function process_request() {
		
		if ( empty( $this->transaction_id ) ) throw new Exception( "transaction_id missing" );
		if ( empty( $this->merchant_id ) ) throw new Exception( "merchant_id missing" );
		if ( empty( $this->amount ) ) throw new Exception( "amount missing" );
		if ( empty( $this->currency ) ) throw new Exception( "currency missing" );

		$this->generate_hmac_signature( '', $this->transaction_id, $this->merchant_id, $this->amount_in_cents, $this->currency );
		$this->generate_blowfish_data();
		$this->generate_form_action();
	}


	public function generate_hmac_signature( ...$parameters ) {
		$sha_string = join( '*', $parameters );

		if ( empty( $this->hmac_key ) ) throw new Exception( "hmac_key missing" );
		
		$this->hmac_signature = strtoupper( hash_hmac( 'sha256', $sha_string, $this->hmac_key ) );
	}


	public function generate_blowfish_data() {
		$data = array();
		
		$this->blowfish_parameters[ 'MsgVer' ] = $this->message_version;
		$this->blowfish_parameters[ 'Response' ] = $this->response;
		
		foreach ( $this->blowfish_mandatory_request_parameters as $key) {
			if ( ! isset( $this->blowfish_parameters[ $key ] ) ) throw new Exception( "$key missing" );
		}
		foreach ( $this->blowfish_parameters as $key => $value ) {
			$data[] = $key . '=' . $value;
		}
		if ( empty( $this->hmac_signature ) ) throw new Exception( "hmac_signature missing" );
		$data[] = 'MAC=' . $this->hmac_signature;
		
		$blowfish_string = join( '&', $data );
		$this->blowfish_length = strlen( $blowfish_string );
		$this->blowfish_data = bin2hex( $this->blowfish_encrypt( $blowfish_string, $this->blowfish_key ) );
	}


	public function generate_form_action() {
		$args = array(
			'CustomField1' => $this->amount . ' ' . $this->currency,
		);
		if ( !empty( $this->language ) ) $args[ 'Language' ] = $this->language;

		$this->form_action = add_query_arg( $args, $this->form_action );
	}


	public function get_form_action() {
		return $this->form_action;
	}


	public function get_merchant_id() {
		return $this->merchant_id;
	}


	public function get_blowfish_data() {
		return $this->blowfish_data;
	}


	public function get_blowfish_length() {
		return $this->blowfish_length;
	}


	private function blowfish_encrypt(string $data, string $key) {
		$l = strlen($key);
		if ($l < 16) {
			$key = str_repeat($key, (int) ceil(16 / $l));
		}
	
		if (($m = strlen($data) % 8) > 0) {
			$data .= str_repeat("\x00", 8 - $m);
		}
	
		$val = openssl_encrypt($data, 'BF-ECB', $key, OPENSSL_RAW_DATA | OPENSSL_NO_PADDING);
	
		return (string) $val;
	}


	private function blowfish_decrypt(string $data, string $key) {
		$l = strlen($key);
		if ($l < 16) {
			$key = str_repeat($key, (int) ceil(16 / $l));
		}
	
		$val = openssl_decrypt($data, 'BF-ECB', $key, OPENSSL_RAW_DATA | OPENSSL_NO_PADDING);
	
		return rtrim((string) $val, "\0");
	}


	public function process_response( $http_request ) {
		
		$hmac_parameters = array();
		
		$data = $http_request[ 'Data' ];
		$length = $http_request[ 'Len' ];
		
		$plain_text = $this->blowfish_decrypt( (string) hex2bin( $data ), $this->blowfish_key );

		$a = explode( '&', $plain_text );
		
		if ( is_array( $a ) and count( $a ) ) {
			foreach ( $a as $string ) {
				if ( preg_match( '/=/', $string ) ) {
					list( $key, $value ) = explode( '=', $string, 2 );
					if ( $key == 'MIB' ) $key = 'mib';
					$this->response_parameters[ $key ] = (string) $value;
					if ( in_array( $key, $this->blowfish_mandatory_response_parameters ) ) {
						$hmac_parameters[ $key ] = (string) $value;
					}
				}
			}
		}
		
		foreach ( $this->blowfish_mandatory_response_parameters as $key ) {
			if ( ! isset( $hmac_parameters[ $key ] ) ) {
				throw new Exception( "$key missing " . print_r( $hmac_parameters, true ) . print_r( $this->blowfish_mandatory_response_parameters, true ) . print_r( $this->response_parameters, true ) );
			}
		}
		
		if ( ! isset( $this->response_parameters[ 'MAC' ] ) and empty( $this->response_parameters[ 'MAC' ] ) ) {
			throw new Exception( "MAC missing" );
		}
		
		$this->generate_hmac_signature( $hmac_parameters[ 'PayID' ], $hmac_parameters[ 'TransID' ], $hmac_parameters[ 'mid' ], $hmac_parameters[ 'Status' ], $hmac_parameters[ 'Code' ] );
		
		if ( strcmp( $this->hmac_signature, trim( $this->response_parameters[ 'MAC' ] ) ) != 0 ) {
			throw new Exception( 'Wrong MAC signature' );
			$this->response_is_valid = false;
		}
		
		$this->response_is_valid = true;
		
		$this->response_is_successful = ($this->response_parameters[ 'Code' ] === '00000000');
	}

	
	public function get_response_parameter( $key ) {
		if ( empty( $this->response_parameters ) ) {
			throw new Exception( "Response parameters empty" );
		}
		if ( ! isset( $this->response_parameters[ $key ] ) or empty( $this->response_parameters[ $key ] ) ) {
			throw new Exception( "Response parameter $key not set or empty" );
		}
		return $this->response_parameters[ $key ];
	}


	public function get_response_parameters() {
		if ( empty( $this->response_parameters ) ) {
			throw new Exception( "Response parameters empty" );
		}
		return $this->response_parameters;
	}


	public function response_is_valid() {
		return $this->response_is_valid;
	}


	public function response_is_successful() {
		if ( ! $this->response_is_valid ) {
			throw new Exception( "Response is not valid, cannot call response_is_successful()" );
		}
		return $this->response_is_successful;
	}
}



includes/class-axepta-updater.php
<?php


namespace Axepta;


if ( ! defined( 'ABSPATH' ) ) {
	exit;
}


class Updater {

	private $file;
	private $server_url;
	private $plugin;
	private $basename;
	private $active;
	private $server_response;

	public function __construct( $file, $server_url ) {

		$this->file = $file;
		$this->server_url = $server_url;

		add_action( 'admin_init', array( $this, 'set_plugin_properties' ) );

		return $this;
	}

	public function set_plugin_properties() {
		$this->plugin	= get_plugin_data( $this->file );
		$this->basename = plugin_basename( $this->file );
		$this->active	= is_plugin_active( $this->basename );
	}

	private function get_server_info() {
		$this->set_plugin_properties();
		
		if ( is_null( $this->server_response ) ) {
			$slug = current( explode( '/' , $this->basename ) );
			$request_uri = $this->server_url.'/'.$slug.'.json';
			
			$response = json_decode( wp_remote_retrieve_body( wp_remote_get( $request_uri ) ), true ); // Get JSON and parse it

			$this->server_response = $response;
		}
	}

	public function initialize() {
		add_filter( 'pre_set_site_transient_update_plugins', array( $this, 'modify_transient' ), 10, 1 );
		add_filter( 'plugins_api', array( $this, 'plugin_popup' ), 10, 3);
		add_filter( 'upgrader_post_install', array( $this, 'after_install' ), 10, 3 );
	}

	public function modify_transient( $transient ) {

		if ( property_exists( $transient, 'checked') ) {

			if ( $checked = $transient->checked ) {

				$this->get_server_info();
				
				
				$out_of_date = version_compare( $this->server_response['version'], $checked[ $this->basename ], 'gt' ); // Check if we're out of date

				if ( $out_of_date ) {
					
					$slug = current( explode( '/' , $this->basename ) );
					
					$plugin = array(
						'url' => $this->plugin["PluginURI"],
						'slug' => $slug,
						'package' => $this->server_response['download_url'],
						'new_version' => $this->server_response['version']
					);
					
					$transient->response[$this->basename] = (object) $plugin;

				}
			}
		}

		return $transient;
	}

	public function plugin_popup( $result, $action, $args ) {

		if ( ! empty( $args->slug ) ) {
			
			if ( $args->slug == current( explode( '/' , $this->basename ) ) ) {
				
				$this->get_server_info();
				
				return (object) $this->server_response;

			}

		}
		
		return $result;
	}
	
	public function after_install( $response, $hook_extra, $result ) {
		global $wp_filesystem; // Get global FS object

		$install_directory = plugin_dir_path( $this->file ); // Our plugin directory
		$wp_filesystem->move( $result['destination'], $install_directory ); // Move files to the plugin dir
		$result['destination'] = $install_directory; // Set the destination for the rest of the stack

		if ( $this->active ) { // If it was active
			activate_plugin( $this->basename ); // Reactivate
		}

		return $result;
	}

}



assets/js/woocommerce-gateway-axepta.js
jQuery(document).ready(function() {
	jQuery('#woocommerce_axepta_gateway_test_mode').on('change', function() {
		if (jQuery(this).is(':checked')) {
			jQuery('#woocommerce_axepta_gateway_merchant_id, #woocommerce_axepta_gateway_hmac_key, #woocommerce_axepta_gateway_blowfish_key').parents('tr').hide();
			jQuery('#woocommerce_axepta_gateway_test_merchant_id, #woocommerce_axepta_gateway_test_hmac_key, #woocommerce_axepta_gateway_test_blowfish_key').parents('tr').show();
		} else {
			jQuery('#woocommerce_axepta_gateway_test_merchant_id, #woocommerce_axepta_gateway_test_hmac_key, #woocommerce_axepta_gateway_test_blowfish_key').parents('tr').hide();
			jQuery('#woocommerce_axepta_gateway_merchant_id, #woocommerce_axepta_gateway_hmac_key, #woocommerce_axepta_gateway_blowfish_key').parents('tr').show();
		}
	});
	jQuery('#woocommerce_axepta_gateway_test_mode').trigger('change');
});


woocommerce-gateway-axepta.php
<?php

/**
 * Plugin Name: WooCommerce Axepta Gateway
 * Description: Accept card payments on your store using Axepta's gateway.
 * Plugin URI: https://github.com/studiomaiis/wp-plugin-woocommerce-gateway-axepta
 * Version: 0.3
 * Requires at least: 5.6.8
 * Tested up to: 6.0.2
 * Requires PHP: 7.3
 * Author: Pierre BASSON
 * Author URI: https://www.studiomaiis.net
 * Text Domain: woocommerce-gateway-axepta
 * Domain Path: /languages
 *
 * WC requires at least: 5.8
 * WC tested up to: 6.9.4
 *
 * License: GNU General Public License v3.0
 * License URI: https://www.gnu.org/licenses/gpl-3.0.html
 */


if ( ! defined( 'ABSPATH' ) ) {
	exit;
}


/**
 * Required minimums and constants
 */
define( 'WC_AXEPTA_SERVER_URL', 'https://depot.studiomaiis.net/wordpress' );
define( 'WC_AXEPTA_PREFIX', 'wc_axepta' );
define( 'WC_AXEPTA_PLUGIN_NAME', __( 'WooCommerce Axepta Gateway', 'woocommerce-gateway-axepta') );
define( 'WC_AXEPTA_MAIN_FILE', __FILE__ );
define( 'WC_AXEPTA_ABSPATH', __DIR__ . '/' );
define( 'WC_AXEPTA_PLUGIN_URL', untrailingslashit( plugins_url( basename( plugin_dir_path( __FILE__ ) ), basename( __FILE__ ) ) ) );
define( 'WC_AXEPTA_PLUGIN_PATH', untrailingslashit( plugin_dir_path( __FILE__ ) ) );


function woocommerce_gateway_axepta() {

	static $plugin;
	
	if ( ! isset( $plugin ) ) {
	
		class WC_Axepta {
	
			/**
			 * The *Singleton* instance of this class
			 *
			 * @var Singleton
			 */
			private static $instance;


			/**
			 * Returns the *Singleton* instance of this class.
			 *
			 * @return Singleton The *Singleton* instance.
			 */
			public static function get_instance() {
				if ( null === self::$instance ) {
					self::$instance = new self();
				}
				return self::$instance;
			}


			/**
			 * The main Axepta gateway instance. Use get_main_axepta_gateway() to access it.
			 *
			 * @var null|WC_Axepta_Payment_Gateway
			 */
			protected $gateway = null;


			/**
			 * Private clone method to prevent cloning of the instance of the
			 * *Singleton* instance.
			 *
			 * @return void
			 */
			public function __clone() {}


			/**
			 * Private unserialize method to prevent unserializing of the *Singleton*
			 * instance.
			 *
			 * @return void
			 */
			public function __wakeup() {}


			/**
			 * Protected constructor to prevent creating a new instance of the
			 * *Singleton* via the `new` operator from outside of this class.
			 */
			public function __construct() {
				add_action( 'admin_init', [ $this, 'install' ] );
				
				$this->init();
			}


			/**
			 * Init the plugin after plugins_loaded so environment variables are set.
			 */
			public function init() {
				require_once dirname( __FILE__ ) . '/includes/class-wc-gateway-axepta.php';
				require_once dirname( __FILE__ ) . '/includes/class-axepta-updater.php';

				add_filter( 'woocommerce_payment_gateways', [ $this, 'add_gateways' ] );
				add_filter( 'plugin_action_links_' . plugin_basename( __FILE__ ), [ $this, 'plugin_action_links' ] );
				
				$updater = new Axepta\Updater( __FILE__, WC_AXEPTA_SERVER_URL );
				$updater->initialize();
			}


			/**
			 * Updates the plugin version in db
			 */
			public function update_plugin_version() {
				delete_option( WC_AXEPTA_PREFIX . '_version' );

				if( ! function_exists('get_plugin_data') ){
					require_once( ABSPATH . 'wp-admin/includes/plugin.php' );
				}
				$plugin_data = get_plugin_data( __FILE__ );

				update_option( WC_AXEPTA_PREFIX . '_version', $plugin_data['Version'] );
			}


			/**
			 * Handles upgrade routines.
			 */
			public function install() {
				if ( ! is_plugin_active( plugin_basename( __FILE__ ) ) ) {
					return;
				}
	
				$this->update_plugin_version();
			}


			/**
			 * Add plugin action links.
			 */
			public function plugin_action_links( $links ) {
				$plugin_links = array(
					'<a href="admin.php?page=wc-settings&tab=checkout&section=axepta_gateway">' . esc_html__( 'Settings', 'woocommerce' ) . '</a>',
				);
				return array_merge( $plugin_links, $links );
			}


			/**
			 * Add the gateways to WooCommerce.
			 */
			public function add_gateways( $methods ) {
				$methods[] = $this->get_main_axepta_gateway();
	
				return $methods;
			}


			/**
			 * Returns the main Axepta payment gateway class instance.
			 *
			 * @return WC_Gateway_Axepta
			 */
			public function get_main_axepta_gateway() {
				if ( ! is_null( $this->gateway ) ) {
					return $this->gateway;
				}
	
				$this->gateway = new WC_Gateway_Axepta();
	
				return $this->gateway;
			}
		}
	
		$plugin = WC_Axepta::get_instance();
	
	}
	
	return $plugin;
}


add_action( 'plugins_loaded', 'woocommerce_gateway_axepta_init' );
function woocommerce_gateway_axepta_init() {
	load_plugin_textdomain( 'woocommerce-gateway-axepta', false, dirname( plugin_basename( __FILE__ ) ) . '/languages' );

	if ( ! class_exists( 'WooCommerce' ) ) {
		add_action( 'admin_notices', 'woocommerce_axepta_missing_wc_notice' );
		return;
	}

	woocommerce_gateway_axepta();
}


function woocommerce_axepta_missing_wc_notice() {
	/* translators: 1. Plugin name 2. URL link. */
	echo '<div class="error"><p><strong>' . sprintf( esc_html__( '%1$s requires WooCommerce to be installed and active. You can download %2$s here.', 'woocommerce-gateway-axepta' ), WC_AXEPTA_PLUGIN_NAME, '<a href="https://woocommerce.com/" target="_blank">WooCommerce</a>' ) . '</strong></p></div>';
}


add_action( 'parse_request', 'woocommerce_axepta_parse_request' );
function woocommerce_axepta_parse_request( &$wp ) {
	if ( isset( $wp->query_vars[ 'pagename' ] ) ) {
		if ( isset( $wp->query_vars[ 'order-received' ] ) ) {
			if ( preg_match( '/^axepta:([^:]+):([^:]+)$/', $wp->query_vars[ 'order-received' ], $matches ) ) {
				$order_received_url = wc_get_endpoint_url( 'order-received', $matches[ 1 ], wc_get_checkout_url() );
				$order_received_url = add_query_arg( 'key', $matches[ 2 ], $order_received_url );
				wp_redirect( $order_received_url );
				exit;
			}
		}
	}
}

